### PR TITLE
nrf async twim

### PR DESCRIPTION
**Pull request by folkertdev**
firstly, I have very little idea of what i'm doing here, trying to puzzle things together.

This PR adds async twim (by implementing the I2c trait) for embassy-nrf. This is based on the nrf-hal version of twim, and the embassy async spim implementation.

### CODE REVIEW

File: embassy-nrf/src/twim.rs [hunk 1]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+    #[rustfmt::skip]
+    type WriteReadFuture<'a> where Self: 'a = impl Future<Output = Result<(), Self::Error>> + 'a;
+
+    fn read<'a>(&'a mut self, address: u8, buffer: &'a mut [u8]) -> Self::ReadFuture<'a> {
+        self.write_read(address, &[], buffer)
+    }
+
+    fn write<'a>(&'a mut self, address: u8, bytes: &'a [u8]) -> Self::WriteFuture<'a> {
+        self.write_read(address, bytes, &mut [])
+    }
```
Comment (folkertdev) [inline embassy-nrf/src/twim.rs:464]: is this valid for I2c like it is for SPI?

File: embassy-nrf/src/twim.rs [hunk 2]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+    fn write_read<'a>(
+        &'a mut self,
+        address: u8,
+        bytes: &'a [u8],
+        buffer: &'a mut [u8],
+    ) -> Self::WriteReadFuture<'a> {
+        async move {
+            slice_in_ram_or(bytes, Error::DMABufferNotInDataMemory)?;
+            // NOTE: RAM slice check for buffer is not necessary, as a mutable
+            // slice can only be built from data located in RAM.
```
Comment (folkertdev) [inline embassy-nrf/src/twim.rs:475]: this check is not skipped in the embassy spim implementation, but I think skipping it is safe for the reasons in the comment. Maybe this should also be changed in the spim implementation?

File: embassy-nrf/src/twim.rs [hunk 3]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+
+            // Reset and enable the events
+            r.events_stopped.reset();
+            r.events_error.reset();
+            r.events_lasttx.reset();
+            self.clear_errorsrc();
+
+            r.intenset.write(|w| w.stopped().set());
+            r.intenset.write(|w| w.error().set());
+            r.intenset.write(|w| w.lasttx().set());
```
Comment (folkertdev) [inline embassy-nrf/src/twim.rs:502]: Here I'm not sure. We definitely want to listen to `stopped`, I think we'd also need to know about `error`, but is `lasttx` important too?

File: embassy-nrf/src/twim.rs [hunk 4]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+                    r.events_stopped.reset();
+
+                    return Poll::Ready(());
+                }
+
+                // stop if an error occured
+                if r.events_error.read().bits() != 0 {
+                    r.events_error.reset();
+                    r.tasks_stop.write(|w| unsafe { w.bits(1) });
+                }
```
Comment (folkertdev) [inline embassy-nrf/src/twim.rs:531]: would a loop after this block waiting for `events_stopped` to be set be preferable here?

File: embassy-traits/src/i2c.rs [hunk 5]
```diff
@@ -94,9 +94,15 @@ pub trait I2c<A: AddressMode = SevenBitAddress> {
...
-    type WriteReadFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a;
+    type WriteFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+    type ReadFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+    type WriteReadFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
```
Comment (folkertdev) [inline embassy-traits/src/i2c.rs:105]: these changes are in line with the `FullDuplex` trait.

File: embassy-traits/src/i2c.rs [hunk 6]
```diff
@@ -116,7 +122,7 @@ pub trait I2c<A: AddressMode = SevenBitAddress> {
     /// - `MAK` = master acknowledge
     /// - `NMAK` = master no acknowledge
     /// - `SP` = stop condition
-    fn read<'a>(&'a mut self, address: A, buffer: &mut [u8]) -> Self::ReadFuture<'a>;
+    fn read<'a>(&'a mut self, address: A, buffer: &'a mut [u8]) -> Self::ReadFuture<'a>;
```
Comment (folkertdev) [inline embassy-traits/src/i2c.rs:125]: likewise, the addition of a lifetime on the slice arguments is in line with the `FullDuplex` trait

**Review by folkertdev — COMMENTED** [review]
added some questions/notes

File: embassy-nrf/src/twim.rs (hunk 3 above)
Comment (Dirbaio) [reply embassy-nrf/src/twim.rs:502]: It's slightly faster to do just 1 reg write: 

            r.intenset.write(|w| w.stopped().set().error().set());

File: embassy-nrf/src/twim.rs (hunk 4 above)
Comment (Dirbaio) [reply embassy-nrf/src/twim.rs:531]: No, blocking/spinning is to be avoided if possible, and in this case it definitely is! :) 

It'll keep sleeping and then return Ready when the stopped event happens.

File: embassy-nrf/src/twim.rs [hunk 7]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+
+                Poll::Pending
+            })
+            .await;
+
```
Comment (Dirbaio) [inline embassy-nrf/src/twim.rs:536]: Check errorsrc here, like the blocking code.

File: embassy-nrf/src/twim.rs [hunk 8]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+            let bad_write = r.txd.amount.read().bits() != bytes.len() as u32;
+            let bad_read = r.rxd.amount.read().bits() != buffer.len() as u32;
+
+            if bad_write {
+                return Err(Error::TxBufferTooLong);
```
Comment (Dirbaio) [inline embassy-nrf/src/twim.rs:541]: This should be Error::Transmit, like the blocking code.

File: embassy-nrf/src/twim.rs [hunk 9]
```diff
@@ -437,6 +442,114 @@ impl<'a, T: Instance> Drop for Twim<'a, T> {
...
+            }
+
+            if bad_read {
+                return Err(Error::RxBufferTooLong);
```
Comment (Dirbaio) [inline embassy-nrf/src/twim.rs:545]: This should be Error::Receive, like the blocking code.

File: embassy-traits/src/i2c.rs (hunk 5 above)
Comment (Dirbaio) [reply embassy-traits/src/i2c.rs:105]: Yup, this is needed, thanks for adding it!:

File: embassy-traits/src/i2c.rs (hunk 6 above)
Comment (Dirbaio) [reply embassy-traits/src/i2c.rs:125]: Same :+1:

**Review by Dirbaio — CHANGES_REQUESTED** [review]
Looks good overall! Minor nits regarding error handling, and also `read` and `write` should get their own impls, they can't be implemented in terms of `write_read` because they do different things.

**Comment by folkertdev** [comment]
I think I fixed all things you pointed out (thanks for that btw, very helpful!)

**Review by Dirbaio — APPROVED** [review]
Looks great now! Thanks for the contribution! :)

**Pull request merged**